rancher/rancher
# Omit GroupPrincipals if field is empty

Added omitempty to the defintion of the GroupPrincipals field in the Token struct.  This will prevent that field from showing GroupPrincipals: null when a token is created without GroupPrincipals set.

## Issue: 

https://github.com/rancher/rancher/issues/39107
 
## Problem

Currently, it is possible for a Token to be generated with the GroupPrincipals field set to null in the yaml.  When someone tries to update that token, the effort will fail since null is not a valid value.  To make their update stick, they would need to change GroupPrincipals to be [] (or another valid value) as part of their update.

## Solution

The GroupPrincipals field is now set to "omitempty" in the definition of the Token struct, so it will just not be present if it is not set rather than defaulting it to null.
 
## Testing


## Engineering Testing
### Manual Testing

I've ran many clusters and created many tokens with this update.  In all cases, the GroupPrincipals field is no longer included in the object and all of the created tokens can be updated without incident.

## QA Testing Considerations

1. Install Rancher
2. Verify that all tokens do NOT include `GroupPrincipals: null` in the yaml.  Unless you created a token that specifically has GroupPrincipals, that field will NOT be present at all in the token object.

 
### Regressions Considerations
